src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4345,32 +4578,101 @@ private uint GetTupleValEscape(ImmutableArray<BoundExpression> elements, uint sc
             return narrowestScope;
         }
 
+        /// <summary>
+        /// The escape value of an object initializer is calculated by looking at all of the
+        /// expressions that can be stored into the implicit receiver. That means arguments
+        /// passed to an indexer for example only matter if they can escape into the receiver
+        /// as a stored field.
+        /// </summary>
         private uint GetValEscapeOfObjectInitializer(BoundObjectInitializerExpression initExpr, uint scopeOfTheContainingExpression)
         {
             var result = CallingMethodScope;
-            foreach (var expression in initExpr.Initializers)
+            foreach (var expr in initExpr.Initializers)
             {
-                if (expression.Kind == BoundKind.AssignmentOperator)
-                {
-                    var assignment = (BoundAssignmentOperator)expression;
-                    var rightValEscape = assignment.IsRef
-                        ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
-                        : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
+                var exprResult = GetValEscapeOfObjectMemberInitializer(expr, scopeOfTheContainingExpression);
+                result = Math.Max(result, exprResult);
+            }
 
-                    result = Math.Max(result, rightValEscape);
+            return result;
+        }
+
+        private uint GetValEscapeOfObjectMemberInitializer(BoundExpression expr, uint scopeOfTheContainingExpression)
+        {
+            uint result;
+            if (expr.Kind == BoundKind.AssignmentOperator)
+            {
+                var assignment = (BoundAssignmentOperator)expr;
+                var rightEscape = assignment.IsRef
+                    ? GetRefEscape(assignment.Right, scopeOfTheContainingExpression)
+                    : GetValEscape(assignment.Right, scopeOfTheContainingExpression);
 
-                    var left = (BoundObjectInitializerMember)assignment.Left;
-                    result = Math.Max(result, GetValEscape(left.Arguments, scopeOfTheContainingExpression));
+                uint leftEscape;
+                var left = (BoundObjectInitializerMember)assignment.Left;
+                if (left.MemberSymbol is PropertySymbol { IsIndexer: true } property)
+                {
+                    Debug.Assert(left.AccessorKind != AccessorKind.Unknown);
+                    MethodInfo methodInfo = MethodInfo.Create(property, left.AccessorKind);
+                    leftEscape = getIndexerEscape(methodInfo, left, rightEscape);
```

Comment:
>leftEscape [](http://example.com/codeflow?start=20&length=10)

Consider assigning `result` here instead, and if we need `leftEscape` for the `else` branch, consider moving the declaration to that branch.